File: python/src/seofleet/checks/geo/link_header.py
"""Ported from src/checks/geo/link-header.ts."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional

from ...types import Check, CheckContext, CheckResult

_ID = "link-header"
_NAME = "Link header (RFC 8288)"
_CATEGORY = "geo"


@dataclass
class ParsedLink:
    url: str
    rel: Optional[str] = None
    params: Dict[str, str] = field(default_factory=dict)


def _split_top_level(header: str) -> List[str]:
    """Splits a raw header value on top-level commas -- i.e. commas that
    are neither inside a `<...>` URI-reference nor inside a `"..."` quoted
    param value. RFC 8288 link-values themselves can contain commas in a
    quoted title param (`title="a, b"`), so a naive `split(",")` would
    misparse a multi-link header in that case."""
    parts: List[str] = []
    depth = 0
    in_quotes = False
    current = ""
    for ch in header:
        if ch == '"':
            in_quotes = not in_quotes
        if not in_quotes:
            if ch == "<":
                depth += 1
            if ch == ">":
                depth = max(0, depth - 1)
        if ch == "," and depth == 0 and not in_quotes:
            parts.append(current)
            current = ""
        else:
            current += ch
    if current.strip():
        parts.append(current)
    return parts


def parse_link_header(header: str) -> List[ParsedLink]:
    """Parses a raw RFC 8288 Link header value into one entry per
    link-value, e.g. `<https://example.com/feed>; rel="alternate"` becomes
    `ParsedLink(url="https://example.com/feed", rel="alternate", params={"rel": "alternate"})`.
    Entries that don't match the `<url>; params` shape are dropped rather
    than raising -- a malformed header is data for the check to report on,
    not a reason to crash the run."""
    links: List[ParsedLink] = []
    for raw_part in _split_top_level(header):
        part = raw_part.strip()
        if not part:
            continue
        if not (part.startswith("<") and ">" in part):
            continue
        end = part.index(">")
        url = part[1:end]
        params_raw = part[end + 1 :]

        params: Dict[str, str] = {}
        for raw_param in params_raw.split(";"):
            param_part = raw_param.strip()
            if not param_part:
                continue
            eq = param_part.find("=")
            if eq == -1:
                continue
            key = param_part[:eq].strip().lower()
            value = param_part[eq + 1 :].strip()
            if value.startswith('"') and value.endswith('"'):
                value = value[1:-1]
            params[key] = value

        links.append(ParsedLink(url=url, rel=params.get("rel"), params=params))
    return links


def _run(ctx: CheckContext) -> CheckResult:
    header = ctx.resources.homepage.link_header

    # An RFC 8288 Link header is machine-readable service discovery
    # (alternate feeds, API entry points, rel="describedby" schemas, etc.)
    # that almost no site sends today -- its absence is informational only
    # and never fails the run, same spirit as llms_txt.py.
    if not header or not header.strip():
        return CheckResult(
            _ID, _NAME, _CATEGORY, "WARN",
            "The homepage does not send a Link response header.",
            'Optional: add an RFC 8288 Link response header (e.g. <https://example.com/feed>; rel="alternate") '
            "to advertise machine-readable service-discovery endpoints to crawlers and AI agents.",
        )

    links = parse_link_header(header)
    if not links:
        return CheckResult(
            _ID, _NAME, _CATEGORY, "WARN",
            f'The homepage sends a Link header that could not be parsed as RFC 8288: "{header}".',
            'Optional: format the Link header per RFC 8288, e.g. <https://example.com/feed>; rel="alternate".',
        )

    summary = ", ".join(f'{link.url} (rel="{link.rel}")' if link.rel else link.url for link in links)
    plural = "y" if len(links) == 1 else "ies"
    return CheckResult(
        _ID, _NAME, _CATEGORY, "PASS",
        f"The homepage sends {len(links)} Link header entr{plural}: {summary}.",
    )


link_header_check = Check(id=_ID, name=_NAME, category=_CATEGORY, run=_run)

File: src/checks/geo/link-header.ts
import type { Check, CheckContext, CheckResult } from "../../types.js";

const ID = "link-header";
const NAME = "Link header (RFC 8288)";
const CATEGORY = "geo" as const;

export interface ParsedLink {
  url: string;
  rel?: string;
  params: Record<string, string>;
}

/**
 * Splits a raw header value on top-level commas -- i.e. commas that are
 * neither inside a `<...>` URI-reference nor inside a `"..."` quoted
 * param value. RFC 8288 link-values themselves can contain commas in a
 * quoted title param (`title="a, b"`), so a naive `split(",")` would
 * misparse a multi-link header in that case.
 */
function splitTopLevel(header: string): string[] {
  const parts: string[] = [];
  let depth = 0;
  let inQuotes = false;
  let current = "";
  for (const ch of header) {
    if (ch === '"') inQuotes = !inQuotes;
    if (!inQuotes) {
      if (ch === "<") depth++;
      if (ch === ">") depth = Math.max(0, depth - 1);
    }
    if (ch === "," && depth === 0 && !inQuotes) {
      parts.push(current);
      current = "";
    } else {
      current += ch;
    }
  }
  if (current.trim()) parts.push(current);
  return parts;
}

/**
 * Parses a raw RFC 8288 Link header value into one entry per link-value,
 * e.g. `<https://example.com/feed>; rel="alternate"` becomes
 * `{ url: "https://example.com/feed", rel: "alternate", params: { rel: "alternate" } }`.
 * Entries that don't match the `<url>; params` shape are dropped rather
 * than throwing -- a malformed header is data for the check to report on,
 * not a reason to crash the run.
 */
export function parseLinkHeader(header: string): ParsedLink[] {
  const links: ParsedLink[] = [];
  for (const rawPart of splitTopLevel(header)) {
    const part = rawPart.trim();
    if (!part) continue;

    const match = /^<([^>]*)>(.*)$/.exec(part);
    if (!match) continue;
    const url = match[1] ?? "";
    const paramsRaw = match[2] ?? "";

    const params: Record<string, string> = {};
    for (const rawParam of paramsRaw.split(";")) {
      const paramPart = rawParam.trim();
      if (!paramPart) continue;
      const eq = paramPart.indexOf("=");
      if (eq === -1) continue;
      const key = paramPart.slice(0, eq).trim().toLowerCase();
      let value = paramPart.slice(eq + 1).trim();
      if (value.startsWith('"') && value.endsWith('"')) value = value.slice(1, -1);
      params[key] = value;
    }

    links.push({ url, ...(params.rel !== undefined ? { rel: params.rel } : {}), params });
  }
  return links;
}

export const linkHeaderCheck: Check = {
  id: ID,
  name: NAME,
  category: CATEGORY,
  run(ctx: CheckContext): CheckResult {
    const base = { id: ID, name: NAME, category: CATEGORY };
    const header = ctx.resources.homepage.linkHeader;

    // An RFC 8288 Link header is machine-readable service discovery
    // (alternate feeds, API entry points, rel="describedby" schemas, etc.)
    // that almost no site sends today -- its absence is informational only
    // and never fails the run, same spirit as llms-txt.ts.
    if (!header || !header.trim()) {
      return {
        ...base,
        status: "WARN",
        message: "The homepage does not send a Link response header.",
        fix: 'Optional: add an RFC 8288 Link response header (e.g. <https://example.com/feed>; rel="alternate") to advertise machine-readable service-discovery endpoints to crawlers and AI agents.',
      };
    }

    const links = parseLinkHeader(header);
    if (links.length === 0) {
      return {
        ...base,
        status: "WARN",
        message: `The homepage sends a Link header that could not be parsed as RFC 8288: "${header}".`,
        fix: 'Optional: format the Link header per RFC 8288, e.g. <https://example.com/feed>; rel="alternate".',
      };
    }

    const summary = links.map((l) => (l.rel ? `${l.url} (rel="${l.rel}")` : l.url)).join(", ");
    return {
      ...base,
      status: "PASS",
      message: `The homepage sends ${links.length} Link header entr${links.length === 1 ? "y" : "ies"}: ${summary}.`,
    };
  },
};
